ipn/ipnlocal/local_test.go

```diff
@@ -1330,3 +1331,283 @@ func (rc *routeCollector) AdvertiseRoute(pfx netip.Prefix) error {
 	rc.routes = append(rc.routes, pfx)
 	return nil
 }
+
+type mockSyspolicyHandler struct {
+	t             *testing.T
+	key           syspolicy.Key
+	s             string
+	exitNodeIDKey bool
+	exitNodeIPKey bool
+	exitNodeID    string
+	exitNodeIP    string
+	bothKeys      bool
+	err           error
+}
+
+func (h *mockSyspolicyHandler) ReadString(key string) (string, error) {
+	if h.exitNodeIDKey && key == string(syspolicy.ExitNodeID) {
+		return h.exitNodeID, h.err
+	}
+	if h.exitNodeIPKey && key == string(syspolicy.ExitNodeIP) {
+		return h.exitNodeIP, h.err
+	}
+	return "", h.err
+}
+
+func (h *mockSyspolicyHandler) ReadUInt64(key string) (uint64, error) {
+	return 0, h.err
+}
+
+func (h *mockSyspolicyHandler) ReadBoolean(key string) (bool, error) {
+	return false, h.err
+}
+
+func TestSetExitNodeIDPolicy(t *testing.T) {
+	pfx := netip.MustParsePrefix
+	tests := []struct {
+		name           string
+		exitNodeIPKey  bool
+		exitNodeIDKey  bool
+		exitNodeID     string
+		exitNodeIP     string
+		prefs          *ipn.Prefs
+		exitNodeIPWant string
+		exitNodeIDWant string
+		prefsChanged   bool
+		nm             *netmap.NetworkMap
+	}{
+		{
+			name:           "ExitNodeID key is set",
+			exitNodeIDKey:  true,
+			exitNodeID:     "123",
+			prefs:          ipn.NewPrefs(),
+			exitNodeIDWant: "123",
+			prefsChanged:   true,
+			nm:             new(netmap.NetworkMap),
```

Comment:
optional: to reduce boilerplate when `nm` or `prefs` are empty, allow `nil` values (so they don't need to be mentioned at all in here) and then in the test loop, make an empty prefs or netmap if `tt.prefs == nil` or `tt.nm == nil` respectively